### ISSUE TITLE
Design of specific backends (GPU, OpenMP, Coarrays, etc.)

### ISSUE THREAD

**Issue by certik** [lcompilers/lpython#1996]
So far most of the code that our backends (C, LLVM, WASM) generate does not depend on any third party API, it just uses native operations (such as arithmetic) of the given platform, sometimes `libc`, and sometimes it calls into our own runtime library.

We now have to figure out how to target backends that make heavy use of a custom 3rd party API (typically C) to do all the operations. Examples of such backends:

* SymEngine symbolic backend
* GPU and other accelerator backends
* OpenMP
* Coarrays
* pthreads

The two approaches are:

* We represent the operations in ASR, either with backend explicit nodes, or with higher level operations (parallel do concurrent). Each backend then to implement translating the operation to specific API calls (say OpenMP or SymEngine). Each backend has to reimplement it.
* We do this translation as ASR->ASR pass. The input is, say, do concurrent, and the output is ASR with specific calls to OpenMP (if OpenMP is used) or GPU API (if GPU offloading is used). All backends work with it.

We can use a combination of the two approaches. But the second approach is preferable, since we can see how the code looks like after the transformation (of "do concurrent" into OpenMP or CUDA) and optionally apply more ASR->ASR passes further optimizing the code. And we can use our verify() to check correctness.

**Comment by rebcabin** [comment]
Yes: delaying commitments to API details to the last possible phase
preserves options in the earlier phases. Once those commitments are made,
all downstream phases are stuck with them :)

**Comment by certik** [comment]
I think we can always write an ASR->ASR pass that implements some feature using some low level API. When we use it, then the backends just work. The downside is a possibly slower compilation speed. Later, we can "fuse" it with the backend (=implement it in the backend directly), as long as it can be done in a maintainable way.
